File: java/VmImplementation2/src/repl/Repl.java
package repl;

import ast.nodes.AstProgram;
import compiler.Bytecode;
import compiler.Compiler;
import environment.Environment;
import evaluator.Evaluator;
import objects.AdIntegerObject;
import objects.AdObject;
import parser.Parser;
import vm.VM;

import java.util.Scanner;

public class Repl {
    private Scanner scanner;
    private Parser parser;
    private AstProgram program;
    private Evaluator evaluator;
    private Compiler compiler;
    private VM vm;

    public Repl() {
        scanner = new Scanner(System.in);
        parser = new Parser();
        program = new AstProgram();
        evaluator = new Evaluator();
        compiler = new Compiler();
        vm = new VM();
    }

    public void loop() {
        String line;
        Environment environment = new Environment();
        while (true) {
            System.out.print(">> ");
            line = scanner.nextLine();
            if (line.equals("exit()")) break;
            program.reset();
            parser.load(line);
            parser.buildStatements(program);
            //evaluator.eval(program, environment);
            compiler.init();
            compiler.compile(program);
            Bytecode bytecode = compiler.getBytecode();
            //compiler.getCode().setInstructions(bytecode.getInstructions());
            //System.out.println(compiler.getCode().print());
            vm.init(bytecode);
            vm.run();
            AdObject stackElem = vm.stackTop();
            System.out.println(((AdIntegerObject) stackElem).getValue());
        }
    }
}

File: cpp/vm/advm2/repl.cpp
#include "repl.h"
#include <iostream>
#include <fstream>

Repl::Repl() {
    program = new ASTProgram();
    gc = new GarbageCollector();
    compiler.gc = gc;
    vm.gc = gc;
}

Repl::~Repl() {
    gc->forceFreeObjects();
    delete gc;
    delete program;
}

void Repl::loop() {
    std::cout << "testing a simple VM\n";
    while(1) {
        std::string line;
        std::cout << ">> ";
        getline(std::cin, line);
        if (line == "exit()") break;
        parser.load(line);
        program->reset();
        parser.buildProgramStatement(program);

        compiler.reset();
        compiler.compile(program);

        Bytecode bytecode = compiler.getBytecode();
        compiler.code.instructions = bytecode.instructions;
        std::cout << compiler.code.toString(); // asta pare ca functioneaza cum trebuie

        vm.load(bytecode);
        vm.run();

        gc->unmarkAllObjects();
        gc->markObjects(vm.stack, vm.sp);
        gc->sweepObjects();

        AdObject* result = vm.last_popped_stack_element();
        if (result != NULL) {
            std::cout << result->inspect() << "\n";
        }
    }
}

void Repl::executeFile(std::ifstream &target) {
    if (target.is_open()) {
        std::string line;
        std::string text;
        while (getline(target ,line)) {
            text += line + "\n";
        }
        parser.load(text);
        program->reset();
        parser.buildProgramStatement(program);

        compiler.reset();
        compiler.compile(program);

        Bytecode bytecode = compiler.getBytecode();
        compiler.code.instructions = bytecode.instructions;
        std::cout << compiler.code.toString(); // asta pare ca functioneaza cum trebuie

        vm.load(bytecode);
        vm.run();

        gc->unmarkAllObjects();
        gc->markObjects(vm.stack, vm.sp);
        gc->sweepObjects();

        AdObject* result = vm.last_popped_stack_element();
        if (result != NULL) {
            std::cout << result->inspect() << "\n";
        }
    }
}